### PR TITLE
feat: Adding API for site, organisation, theme and site configuration creation.

### PR DESCRIPTION
**Pull request by saleem-latif**
**Jira Story:**  [EDE-318](https://edlyio.atlassian.net/browse/EDE-318)

**Description:**
This PR creates a REST endpoint that allows its consumers to create a site, organization, site theme and site configuration in a single API call.

### CODE REVIEW

File: lms/urls.py [hunk 1]
```diff
@@ -143,6 +143,9 @@
 
     url(r'^dashboard/', include('learner_dashboard.urls')),
     url(r'^api/experiments/', include('experiments.urls', namespace='api_experiments')),
+
+    # colaraz urls
```
Comment (HamzaIbnFarooq) [inline lms/urls.py:147]: We usually use "[COLARAZ_CUSTOM]" keyword where ever we do some custom changes. With that, we write any description we want.

File: openedx/features/colaraz_features/api/fields.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+"""
+Custom Fields for colaraz API.
+"""
+from django.utils.translation import gettext_lazy as _
+from rest_framework import serializers
+
+from .validators import DomainValidator
```
Comment (HamzaIbnFarooq) [inline openedx/features/colaraz_features/api/fields.py:7]: Can we try avoiding relative imports?

File: openedx/features/colaraz_features/api/serializers.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    @staticmethod
+    def validate_org_name(value):
+        """
+        Validate that site domain does not already exists.
+        """
+        if Organization.objects.filter(name=value).exists():
+            raise serializers.ValidationError('An organization with name "{}" already exists.'.format(value))
+        return value
+
+    def create(self, validated_data):
```
Comment (HamzaIbnFarooq) [inline openedx/features/colaraz_features/api/serializers.py:43]: We have to create two sites, one for lms and one for studio.
We can do two separate requests to add them separately (but in current scenario that is not handled), or we can do it in one single request as we know what kind of URLs will be made for lms and studio.

Comment (saleem-latif) [reply openedx/features/colaraz_features/api/serializers.py:43]: okay, just to be sure I just need to add `studio` in front of the `site_domain` for studio sight? e.g. if `site_domain` is **example.com** studio will be **studio.example.com**

Comment (HamzaIbnFarooq) [reply openedx/features/colaraz_features/api/serializers.py:43]: As far as I know, colaraz is going to use `courses.example.com` for `lms` and `studio.example.com` for `studio`

Comment (saleem-latif) [reply openedx/features/colaraz_features/api/serializers.py:45]: Done

**Comment by saleem-latif** [comment]
@HamzaIbnFarooq I have incorporated your feedback changes, please take another look

File: openedx/features/colaraz_features/helpers.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        SiteConfiguration.objects.create(
+            site=site,
+            enabled=True,
+            values={
+                'SESSION_COOKIE_DOMAIN': site.domain,
+                'PLATFORM_NAME': platform_name,
+                'platform_name': platform_name,
+                'site_domain': site.domain,
+                'site_name': university_name,
+                'course_org_filters': organizations + [organization.short_name]
```
Comment (HamzaIbnFarooq) [inline openedx/features/colaraz_features/helpers.py:111]: Is there any scenario where we get duplicate values?

File: openedx/features/colaraz_features/api/serializers.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    def validate_site_domain(value):
+        """
+        Validate that site domain does not already exists.
+        """
+        if Site.objects.filter(domain=value).exists():
+            raise serializers.ValidationError('A site with domain "{}" already exists.'.format(value))
+        return value
+
+    @staticmethod
+    def validate_org_name(value):
```
Comment (HamzaIbnFarooq) [inline openedx/features/colaraz_features/api/serializers.py:37]: Don't you think we need to check the organization short name too?

File: openedx/features/colaraz_features/helpers.py (hunk 4 above)
Comment (saleem-latif) [reply openedx/features/colaraz_features/helpers.py]: Done

File: openedx/features/colaraz_features/api/serializers.py (hunk 5 above)
Comment (saleem-latif) [reply openedx/features/colaraz_features/api/serializers.py:37]: Done

File: openedx/features/colaraz_features/helpers.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        domain (str): Domain to identify the site. e.g. example.com, test.example.com etc.
+        name (str): Human readable name for the site.
+
+    Returns:
+        (Pair<Site, Site>): named tuple containing lms, studio sites.
+    """
+    sites = (
+        Site.objects.create(
+            domain='{}.{}'.format(prefix, domain),
+            name=name
```
Comment (HamzaIbnFarooq) [inline openedx/features/colaraz_features/helpers.py:51]:
```suggestion
            name='{}.{}'.format(prefix, name)
```
To distinguish between LMS and Studio sites.

**Pull request merged**